Actions and launchers

Antolius opened the pull request:
Resolve #180 
Resolve #179

I did several things here:

1. Explain how to enable DBus activation for an app (_Launchers_ page)
2. Explain how to declare launching actions (_Launchers_ page)
3. Add reference to launching actions (_Actions_ page)
4. Update examples to register action fro the `startup` method (both _Actions_ and _Launchers_ pages)

As discussed in the comments on the #180 issue, I think this is a minimum change which makes sense. We can't really leave any of these steps out and end up with a working app.

An alternative approach I considered was moving parts of this documentation, adding some more stuff and creating a new _Application lifecycle_ page. This new page could include discussion of DBus activation and differences between `startup`, `activate` and `open` signals. Perhaps we could also include a section on opening files, defining new file types and registering the app as default handler for those. But all that seemed to be too much scope for the 2 issues that I was trying to solve. So I gave up on the _Application lifecycle_ page approach and limited the changes to existing _Launchers_ and _Actions_ pages. If that approach sounds interesting I can try scoping it a bit and open a dedicate issue for the  _Application lifecycle_ page.

Antolius commented on apis/actions.md:
```diff
@@ -119,10 +126,8 @@ var button = new Gtk.Button.from_icon_name ("process-stop", Gtk.IconSize.LARGE_T
 
 {% hint style="info" %}
-If you're having trouble, you can view the full example code [here on GitHub](https://github.com/vala-lang/examples/tree/glib-action)
+If you're having trouble, you can view the full example code [here on GitHub](https://github.com/vala-lang/examples/tree/glib-action).  You can learn more about Actions from GNOME [developer documentation](https://developer.gnome.org/documentation/tutorials/actions.html)
```
I'm not sure if it's OK to link to GNOME's documentation, so I added these references (here and in the _Launchers_ page) in a separate commit so that they can be easily removed.

Antolius replied:
This makes sense! I switched the link (and wording around it) to valadoc reference documentation.

Antolius commented on apis/launchers.md:
```diff
@@ -1,5 +1,5 @@
 ---
-description: 'Adding Badges, Progress Bars, and Quick Lists'
+description: 'Adding Badges, Progress Bars, and launching Actions'
```
This nomenclature was changed back in the #98 but the page description still uses the old naming.

Antolius commented:
I also opened a [mini pull-request](https://github.com/vala-lang/examples/pull/5) to the examples repo in the vala-lang org.

danrabbit reviewed (changes requested):
Thanks again for tackling this! I have a few inline comments and then we'll want to update this to match the example code PR too

Antolius commented on apis/launchers.md:
```diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
+You can use any action defined in the `app` namespace as one of the entry points for your application. They appear in the context menu of your app icon in the Applications Menu and Dock, and are searchable by name from the Applications Menu. Implementing actions is covered in-depth in [the actions section](actions).
 
-Adding actions to a `.desktop` file does not involve writing any code or using any external dependencies, though your app needs a way to distinguish between actions, e.g. with command line flags.
+### D-Bus activation
```
Hm... I'm not sure I understand the scope of the separate PR.

We need DBus activation for launching actions. So more-less all the changes on the Launchers page (except perhaps the updated description) need to go together. And cross-linking between the Actions and Launchers pages only makes sense if Launchers actually uses actions.

The one thing we could publish separately is refactored code examples in the Actions page (the split between `activate` and `startup` methods). But that change alone does not offer new capabilities to developers so I personally don't see it as valuable on its own. I can still extract it into separate PR, if you'd like.

danrabbit replied:
Yeah it would reduce the size of this PR making it easier to re-review launching once we're ready for it and it would still be good advice for anyone writing actions in their apps right now. Plus I think the other style changes to the example .destkop file are still an improvement even without dbus activation and just good future proofing. Sorry for the confusion! I would like to get more of your work merged sooner than later if possible :)

Antolius replied:
Ok, I've split the PR in two parts:

1.  #184, which does not depend on DBus activation, and so is independent from dock/#166
2. #185, for which we'll need to wait for the Dock patch to land

I tried to separate the 2 pull requests as cleanly as possible, so there should be no conflicts between them.

Antolius commented on apis/actions.md:
```diff
@@ -51,31 +53,36 @@ If you compile your app, you can see that it now has a custom HeaderBar with a b
 * Add the action to our `Gtk.Application`'s [`ActionMap`](https://valadoc.org/gio-2.0/GLib.ActionMap.html)
 * Set the "accelerators" \(keyboard shortcuts\) for "app.quit" to `<Control>q` and `<Control>w"`. Notice that the action name is prefixed with `app`; this refers to the `ActionMap` built in to `Gtk.Application`
-* Connect to the `activate` signal of our `SimpleAction` and call `destroy ()` on `main_window`. This must be at the end of `activate ()` because of that reference to `main_window`
+* Connect to the `activate` signal of our `SimpleAction` and call `destroy ()` on `main_window`. This is why we saved `main_window` in a private field.
```
The quit action is also mentioned in the Notifications page, so I updated the example there to reflect these changes.

Antolius commented on apis/launchers.md:
```diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
 ## Actions
 
-Actions are specific functions your app can perform without already being open; think of them as alternate and more specific entry points into your app. Actions appear in the context menu of your app icon in the Applications Menu and Dock, and are searchable by name from the Applications Menu.
+You can use any action defined in the `app` namespace as one of the entry points for your application. They appear in the context menu of your app icon in the Applications Menu and Dock, and are searchable by name from the Applications Menu. Implementing actions is covered in-depth in [the actions section](actions).
```
I reverted the intro back to its original wording. I also pulled the cross-link to the Actions page down into the _Declaring actions_ sub-section, since that's the first time this info is actually useful.

Antolius commented on apis/launchers.md:
````diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
 ```
 
-Then below, add the action itself using the same unique ID:
+You can learn more about D-Bus application activation from GNOME [developer documentation](https://developer.gnome.org/documentation/guidelines/maintainer/integrating.html#d-bus-activation).
````
Removed it.

Antolius commented on apis/launchers.md:
```diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
+### Declaring actions
 
-The `Icon` line is optional and should be an icon which represents the action that will be performed. The `Exec` line is required and should be your app's executable name and any command line argument required to trigger the action.
+When defining a `GLib.Action` that can be used to launch the app make sure to register it inside application's `startup` method:
```
I removed the example code snippet and just linked to Actions page.

Antolius commented on apis/launchers.md:
````diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
+Exec=com.github.myteam.myapp
+```
+
+The `Icon` line is optional and should reference a dedicated icon for the action. The `Exec` line should be specified, but is used only for backwards compatibility in case your app ever runs in an environment without D-Bus activation support. Learn more from the [related Freedesktop specification](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s11.html).
````
Uf, yeah, I tried to be clever here and it came out misleading. I reverted this sentence to its original wording.

Antolius commented on apis/launchers.md:
````diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
+```
 
-Let's take a look at an example of an action that opens a new window of an app:
+Actions must also be declared in a new `Actions` line in your app's `.desktop` file. This line should contain a `;` separated list of unique action names:
````
This is an important point to make, especially now that there's no code snippet with actual action creation on this page. I introduced a warning styled hint to clarify it. I'm not sure if warning is suited for this use-case, I used it to signify importance. 🤔️

Antolius commented on apis/launchers.md:
```diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
+The action name should not include your app's name, as it will always be displayed alongside your app. The action icon should also not be your app icon, as it may be shown in the menu for your app icon, or badged on top of the app icon.
+{% endhint %}
 
-See the [freedesktop.org Additional applications actions section](https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s10.html) for a detailed description of what keys are supported and what they do.
```
Makes sense. I reverted to the original wording and moved it to the end of the page. I just updated the href.

Antolius commented on apis/launchers.md:
````diff
@@ -61,48 +61,80 @@ As you can see, the method `set_progress` takes a `double` value type and is a r
+
+```ini
+[Desktop Action my-action]
+Name=User visible name of my action
````
Got it. Nice rule of thumb. I updated the code snippet.

Antolius commented:
> Thanks again for tackling this! I have a few inline comments and then we'll want to update this to match the example code PR too

I updated the PR according to comments and rebased it with the master branch. I resolved the conversations in cases where I think I addressed the issue (I hope that's OK 😅️).

danrabbit commented on apis/actions.md:
```diff
@@ -51,31 +51,32 @@ If you compile your app, you can see that it now has a custom HeaderBar with a b
+* Connect the `activate` signal of our `SimpleAction` to Application's [quit method](https://valadoc.org/gio-2.0/GLib.Application.quit.html).
 
-and now we can tie the action to the HeaderBar Button by assigning the `action_name` property of our Button:
+{% hint style="info" %}
```
I wonder if we should swap this with the hint on line 88/89 since that hint is more immediately related to this section and a "learn more" hint is probably most useful after we finish here and not before we're done

Antolius replied:
I've switched the 2 info boxes in #184 and #185.

Antolius commented:
I'll close this PR in favor of #184 and #185. :tada:
